File: fair-cut.js
# https://www.hackerrank.com/challenges/fair-cut/problem?isFullScreen=true



function fairCut(n, k, arr) {
    arr.sort((a, b) => a - b);
    if (2 * k > n) k = n - k;
    const s = (n - 2 * k) >> 1;
    const e = s + 2 * k;
    let ans = 0;
    let sum1 = 0, cnt1 = 0;
    let sum2 = 0, cnt2 = 0;
    for (let i = 0; i < n; ++i) {
        const x = arr[i];
        if (s <= i && i < e && (i & 1)) {
            sum1 += x;
            cnt1 += 1;
            ans += cnt2 * x - sum2;
        } else {
            sum2 += x;
            cnt2 += 1;
            ans += cnt1 * x - sum1;
        }
    }
    return ans;
}

File: fair-cut.cpp
# https://www.hackerrank.com/challenges/fair-cut/problem?isFullScreen=true



 long long int min(long long int i, long long int k){
     long long int result = (i > k) ? k : i;
     return result;
 }

long long int fairCut(long long int k, vector<long long int> arr) {
   long long int n=arr.size();
   sort(arr.begin(),arr.end());
   vector<vector<long long int>> dp(n+1,vector<long long int>(k+1,1e18));
   dp[0][0]=0;

   for(long long int i=1;i<=n;i++)
   {
       for(long long int j=0;j<=min(i,k);j++)
       {
           
           dp[i%2][j]=min(dp[(i+1)%2][j]+j*arr[i-1]-(k-j)*arr[i-1],(j>0?dp[(i+1)%2][j-1]+(i-j)*arr[i-1]-(n-k-i+j)*arr[i-1]:(long long)1e18));
       }
   }
 return dp[n%2][k];
}
